### ISSUE TITLE
BNGBlaster Report rx-loss and rx-packets / tx-packets is not always matching

### ISSUE THREAD

**Issue by victorRoemer** [rtbrick/bngblaster#308]
**Describe the bug**

When I fetch the report json after BNGBlaster test it has sometimes a missmatch between rx-loss and the diffrence between rx-packets and tx-packets. 

**To Reproduce**

I start BNGBlaster via Controller API than I wait 60 seconds and stop BNGBlaster. Than I wait 15 seconds before sending a request to get the report.json. All commands are done via Controller API. 

- bngblasterctrl-version: 0.1.2
- bngblaster-version: 0.9.15

JSON config:
```json
{
   "__comment__":"allure.label.story:NewProduct in None",
   "interfaces":{
      "network":[
         {
            "__comment__":"IPv2 Backbone (ViperNet): bestEffort",
            "gateway":"214.0.0.1",
            "address":"214.0.0.4/21",
            "interface":"bb-vulcanus01",
            "address-ipv6":"2003:4:4000::1:4/48",
            "gateway-ipv6":"2003:4:4000::1"
         }
      ],
      "access":[
         {
            "__comment__":"53434F4D00011B9E on Line DEU.A4PLAY.8466C6EE70",
            "interface":"ci-vulcanus01",
            "monkey":false,
            "type":"pppoe",
            "stream-group-id":1,
            "inner-vlan":7,
            "outer-vlan":1124
         },
         {
            "__comment__":"53434F4D00011BBF on Line DEU.A4PLAY.0D0C5974B9",
            "interface":"ci-vulcanus01",
            "monkey":false,
            "type":"pppoe",
            "stream-group-id":2,
            "inner-vlan":7,
            "outer-vlan":1123
         },
         {
            "__comment__":"53434F4D00011C3C on Line DEU.A4PLAY.A5653349BE",
            "interface":"ci-vulcanus01",
            "monkey":false,
            "type":"pppoe",
            "stream-group-id":3,
            "inner-vlan":7,
            "outer-vlan":1122
         }
      ]
   },
   "traffic":{
      "stream-burst-ms":10
   },
   "sessions":{
      "count":3,
      "max-outstanding":40,
      "start-rate":40,
      "stop-rate":40,
      "reconnect":true,
      "monkey-autostart":true
   },
   "pppoe":{
      "reconnect":true,
      "discovery-retry":10,
      "vlan-priority":6
   },
   "streams":[
      {
         "name":"sp_bestEffort",
         "stream-group-id":1,
         "__comment__":"This channel has TOS 0: best effort",
         "direction":"downstream",
         "Mbps":400,
         "length":1460,
         "type":"ipv6",
         "network-interface":"bb-vulcanus01"
      },
      {
         "name":"dp_bestEffort",
         "stream-group-id":2,
         "__comment__":"This channel has TOS 0: best effort",
         "direction":"downstream",
         "Mbps":400,
         "length":1460,
         "type":"ipv6",
         "network-interface":"bb-vulcanus01"
      },
      {
         "name":"tp_bestEffort",
         "stream-group-id":3,
         "__comment__":"This channel has TOS 0: best effort",
         "direction":"downstream",
         "Mbps":400,
         "length":1460,
         "type":"ipv6",
         "network-interface":"bb-vulcanus01"
      }
   ]
}
```
JSON report:
```json
{
   "report":{
      "version":"0.9.14",
      "test-duration":123,
      "sessions":3,
      "sessions-pppoe":3,
      "sessions-ipoe":0,
      "sessions-established":3,
      "sessions-flapped":0,
      "setup-time-ms":50526,
      "setup-rate-cps":0.05938,
      "setup-rate-cps-min":0.05938,
      "setup-rate-cps-avg":0.09511,
      "setup-rate-cps-max":0.1305,
      "dhcp-sessions-established":0,
      "dhcpv6-sessions-established":3,
      "interfaces":[
         {
            "name":"ci-vulcanus01",
            "type":"Interface",
            "tx-packets":74,
            "tx-bytes":3486,
            "tx-polled":0,
            "tx-io-error":0,
            "tx-to-long":0,
            "tx-no-buffer":0,
            "rx-packets":8484137,
            "rx-bytes":12607232321,
            "rx-protocol-error":0,
            "rx-unknown":4,
            "rx-polled":12607232321,
            "rx-io-error":0,
            "rx-no-buffer":0
         },
         {
            "name":"bb-vulcanus01",
            "type":"Interface",
            "tx-packets":9916459,
            "tx-bytes":14616836778,
            "tx-polled":0,
            "tx-io-error":0,
            "tx-to-long":0,
            "tx-no-buffer":0,
            "rx-packets":489,
            "rx-bytes":34820,
            "rx-protocol-error":0,
            "rx-unknown":65,
            "rx-polled":34820,
            "rx-io-error":0,
            "rx-no-buffer":0
         }
      ],
      "network-interfaces":[
         {
            "name":"bb-vulcanus01",
            "tx-packets":9916459,
            "tx-multicast-packets":0,
            "rx-packets":424
         }
      ],
      "access-interfaces":[
         {
            "name":"ci-vulcanus01",
            "tx-packets":74,
            "rx-packets":8484133,
            "rx-multicast-packets":0,
            "rx-multicast-packets-loss":0,
            "protocol-stats":{
               "tx-arp":0,
               "rx-arp":0,
               "tx-padi":18,
               "rx-pado":3,
               "tx-padr":3,
               "rx-pads":3,
               "tx-padt":1,
               "rx-padt":2,
               "tx-lcp":28,
               "rx-lcp":27,
               "tx-pap":3,
               "rx-pap":3,
               "tx-chap":0,
               "rx-chap":0,
               "tx-ipcp":9,
               "rx-ipcp":9,
               "tx-ip6cp":6,
               "rx-ip6cp":6,
               "tx-igmp":0,
               "rx-igmp":5,
               "tx-icmp":0,
               "rx-icmp":0,
               "tx-dhcp":0,
               "rx-dhcp":0,
               "tx-dhcpv6":3,
               "rx-dhcpv6":3,
               "tx-icmpv6":3,
               "rx-icmpv6":11,
               "rx-ipv4-fragmented":0,
               "lcp-echo-timeout":1,
               "lcp-request-timeout":0,
               "ipcp-request-timeout":0,
               "ip6cp-request-timeout":0,
               "pap-timeout":0,
               "chap-timeout":0,
               "dhcp-timeout":0,
               "dhcpv6-timeout":0,
               "icmpv6-rs-timeout":0
            }
         }
      ],
      "traffic-streams":{
         "total-flows":3,
         "verified-flows":3,
         "first-seq-rx-min":28735,
         "first-seq-rx-max":1523619,
         "flow-rx-packet-loss-min":54,
         "flow-rx-packet-loss-max":1294659,
         "flow-rx-delay-us-min":256,
         "flow-rx-delay-us-max":43603
      },
      "streams":[
         {
            "flow-id":1,
            "name":"sp_bestEffort",
            "type":"unicast",
            "sub-type":"ipv6",
            "direction":"downstream",
            "enabled":true,
            "active":false,
            "verified":true,
            "source-address":"2003:4:4000::1:4",
            "source-port":65056,
            "destination-address":"2003:4:3000:92f8:200:ff:fe00:1",
            "destination-port":65056,
            "protocol":"udp",
            "tx-interface":"bb-vulcanus01",
            "tx-interface-state":"Up",
            "rx-interface":"ci-vulcanus01",
            "rx-first-seq":1523619,
            "rx-last-seq":2294421,
            "rx-tos-tc":0,
            "rx-ttl":253,
            "rx-outer-vlan-pbit":0,
            "rx-inner-vlan-pbit":0,
            "rx-len":1486,
            "tx-len":1474,
            "tx-packets":2294421,
            "tx-bytes":3381976554,
            "rx-packets":2294506,
            "rx-bytes":3409635916,
            "rx-loss":0,
            "rx-wrong-order":1523703,
            "rx-delay-us-min":256,
            "rx-delay-us-max":5195,
            "rx-pps":31614,
            "tx-pps":31607,
            "rx-pps-max":34257,
            "tx-pps-max":34253,
            "tx-bps-l2":372709744,
            "rx-bps-l2":375827232,
            "rx-bps-l3":369251520,
            "tx-mbps-l2":372.7,
            "rx-mbps-l2":375.8,
            "rx-mbps-l3":369.3,
            "tx-first-epoch":2445007,
            "rx-first-epoch":2445007,
            "rx-last-epoch":2445074,
            "rx-wrong-session":0,
            "session-id":1,
            "session-version":8,
            "session-traffic":false
         },
         {
            "flow-id":2,
            "name":"dp_bestEffort",
            "type":"unicast",
            "sub-type":"ipv6",
            "direction":"downstream",
            "enabled":true,
            "active":false,
            "verified":true,
            "source-address":"2003:4:4000::1:4",
            "source-port":65056,
            "destination-address":"2003:4:3000:92fa:200:ff:fe00:2",
            "destination-port":65056,
            "protocol":"udp",
            "tx-interface":"bb-vulcanus01",
            "tx-interface-state":"Up",
            "rx-interface":"ci-vulcanus01",
            "rx-first-seq":28735,
            "rx-last-seq":2294420,
            "rx-tos-tc":0,
            "rx-ttl":253,
            "rx-outer-vlan-pbit":0,
            "rx-inner-vlan-pbit":0,
            "rx-len":1486,
            "tx-len":1474,
            "tx-packets":2294420,
            "tx-bytes":3381975080,
            "rx-packets":971189,
            "rx-bytes":1443186854,
            "rx-loss":1294561,
            "rx-wrong-order":64,
            "rx-delay-us-min":256,
            "rx-delay-us-max":43603,
            "rx-pps":31613,
            "tx-pps":31607,
            "rx-pps-max":34255,
            "tx-pps-max":34254,
            "tx-bps-l2":372709744,
            "rx-bps-l2":375815344,
            "rx-bps-l3":369239840,
            "tx-mbps-l2":372.7,
            "rx-mbps-l2":375.8,
            "rx-mbps-l3":369.2,
            "tx-first-epoch":2445007,
            "rx-first-epoch":2445008,
            "rx-last-epoch":2445074,
            "rx-wrong-session":0,
            "session-id":2,
            "session-version":8,
            "session-traffic":false
         },
         {
            "flow-id":3,
            "name":"tp_bestEffort",
            "type":"unicast",
            "sub-type":"ipv6",
            "direction":"downstream",
            "enabled":true,
            "active":false,
            "verified":true,
            "source-address":"2003:4:4000::1:4",
            "source-port":65056,
            "destination-address":"2003:4:3000:92f9:200:ff:fe00:3",
            "destination-port":65056,
            "protocol":"udp",
            "tx-interface":"bb-vulcanus01",
            "tx-interface-state":"Up",
            "rx-interface":"ci-vulcanus01",
            "rx-first-seq":1357315,
            "rx-last-seq":2294421,
            "rx-tos-tc":0,
            "rx-ttl":253,
            "rx-outer-vlan-pbit":0,
            "rx-inner-vlan-pbit":0,
            "rx-len":1486,
            "tx-len":1474,
            "tx-packets":2294421,
            "tx-bytes":3381976554,
            "rx-packets":2294506,
            "rx-bytes":3409635916,
            "rx-loss":0,
            "rx-wrong-order":1357399,
            "rx-delay-us-min":256,
            "rx-delay-us-max":5533,
            "rx-pps":31614,
            "tx-pps":31607,
            "rx-pps-max":34257,
            "tx-pps-max":34254,
            "tx-bps-l2":372709744,
            "rx-bps-l2":375827232,
            "rx-bps-l3":369251520,
            "tx-mbps-l2":372.7,
            "rx-mbps-l2":375.8,
            "rx-mbps-l3":369.3,
            "tx-first-epoch":2445007,
            "rx-first-epoch":2445007,
            "rx-last-epoch":2445074,
            "rx-wrong-session":0,
            "session-id":3,
            "session-version":8,
            "session-traffic":false
         }
      ]
   }
}
```

1. Start BNGBlaster
2. Wait 60 Seconds
3. Stop BNGBlaster
4. Wait 15 Seconds
5. Get Report
6. Analyze result

**Expected behavior**

rx-loss = tx-packets - rx-packets

**Comment by GIC-de** [comment]
The behavior you observed is indeed intended and correct. The difference between TX and RX packets can easily be calculated if needed, but the term `loss` must be distinguished from this.

The BNG Blaster calculates packet loss as follows: If missing packets are detected between two received packets of a flow based on their sequence numbers, this is counted as loss. For example, if you receive a packet with sequence number 10 and the next packet has sequence number 14, the loss counter for that flow is increased by 3.

It is important to note that packets lost before the first packet is received, as well as packets sent at the end but no longer received, are not included in this loss counter. For such cases, there are specific fields that record the first and last received sequence numbers. This distinction is particularly crucial for QoS tests.

For example, if I want to verify in a test that there is no packet loss in a specific traffic class, I do not want loss to be counted simply because the stream began sending packets before forwarding was fully ready. On the other hand, if the goal is to determine how long it takes for traffic to start being received, this can be checked by looking at the first received sequence number.

This also explains why RX + loss is not exactly equal to TX.
